core/services/offchainreporting2/delegate.go

```diff
@@ -0,0 +1,340 @@
+//	//OCR2ContractTransmitterTransmitTimeout() time.Duration
+//	//OCR2DatabaseTimeout() time.Duration
+//	//OCR2DefaultTransactionQueueDepth() uint32
+//	//OCR2KeyBundleID() (string, error)
+//	//OCR2TraceLogging() bool
+//	//OCR2TransmitterAddress() (ethkey.EIP55Address, error)
+//
+//	P2PBootstrapPeers() ([]string, error)
+//	P2PPeerID() p2pkey.PeerID
+//	P2PV2Bootstrappers() []ocrcommontypes.BootstrapperLocator
+//}
+
+type Delegate struct {
+	db                    *sqlx.DB
+	jobORM                job.ORM
+	keyStore              keystore.OCR2
+	pipelineRunner        pipeline.Runner
+	peerWrapper           *ocrcommon.SingletonPeerWrapper
+	monitoringEndpointGen telemetry.MonitoringEndpointGenerator
+	chainSet              evm.ChainSet
+	lggr                  logger.Logger
+}
+
+var _ job.Delegate = (*Delegate)(nil)
+
+func NewDelegate(
+	db *sqlx.DB,
+	jobORM job.ORM,
+	keyStore keystore.OCR2,
+	pipelineRunner pipeline.Runner,
+	peerWrapper *ocrcommon.SingletonPeerWrapper,
+	monitoringEndpointGen telemetry.MonitoringEndpointGenerator,
+	chainSet evm.ChainSet,
+	lggr logger.Logger,
+) *Delegate {
+	return &Delegate{
+		db,
+		jobORM,
+		keyStore,
+		pipelineRunner,
+		peerWrapper,
+		monitoringEndpointGen,
+		chainSet,
+		lggr,
+	}
+}
+
+func (d Delegate) JobType() job.Type {
+	return job.OffchainReporting2
+}
+
+func (Delegate) OnJobCreated(spec job.Job) {}
+func (Delegate) OnJobDeleted(spec job.Job) {}
+
+func (Delegate) AfterJobCreated(spec job.Job)  {}
+func (Delegate) BeforeJobDeleted(spec job.Job) {}
+
+func (d Delegate) ServicesForSpec(jobSpec job.Job) (services []job.Service, err error) {
+	if jobSpec.Offchainreporting2OracleSpec == nil {
+		return nil, errors.Errorf("offchainreporting.Delegate expects an *job.Offchainreporting2OracleSpec to be present, got %v", jobSpec)
+	}
+	spec := jobSpec.Offchainreporting2OracleSpec
+
+	chain, err := d.chainSet.Get(jobSpec.Offchainreporting2OracleSpec.EVMChainID.ToInt())
+	if err != nil {
+		return nil, err
+	}
+
+	contract, err := offchain_aggregator_wrapper.NewOffchainAggregator(spec.ContractAddress.Address(), chain.Client())
+	if err != nil {
+		return nil, errors.Wrap(err, "could not instantiate NewOffchainAggregator")
+	}
+
+	contractFilterer, err := ocr2aggregator.NewOCR2AggregatorFilterer(spec.ContractAddress.Address(), chain.Client())
+	if err != nil {
+		return nil, errors.Wrap(err, "could not instantiate NewOffchainAggregatorFilterer")
+	}
+
+	contractCaller, err := ocr2aggregator.NewOCR2AggregatorCaller(spec.ContractAddress.Address(), chain.Client())
+	if err != nil {
+		return nil, errors.Wrap(err, "could not instantiate NewOffchainAggregatorCaller")
+	}
+
+	ocrdb := NewDB(d.db.DB, spec.ID)
+
+	tracker := NewOCRContractTracker(
+		contract,
+		contractFilterer,
+		contractCaller,
+		chain.Client(),
+		chain.LogBroadcaster(),
+		jobSpec.ID,
+		d.lggr,
+		d.db,
+		ocrdb,
+		chain.Config(),
+		chain.HeadBroadcaster(),
+	)
+	services = append(services, tracker)
+
+	var peerID p2pkey.PeerID
```

Comment:
Theres a lot of clean up to be had here pulling out helper functions etc. I have some of it in the ccip repo planning to port that when we add the ocr2-relay job in a follow up

Reply:
yes

Reply:
I'm ok to merge as-is but can we leave a comment with a link to the card to clean it up?